src/replay/replay_engine.rs

```diff
@@ -0,0 +1,174 @@
+    config::{Config, Replay},
+    metrics::REQUEST_DROPPED,
+    replay::{
+        parser::CommandLogLine,
+        replay_speed::{RateController, SpeedController, TimingController},
+    },
+    workload::{
+        client::{Delete, Get, Set},
+        ClientRequest, ClientWorkItemKind,
+    },
+    RUNNING,
+};
+
+pub struct ReplayEngine {
+    command_log_path: String,
+    timing_controller: TimingController,
+}
+
+impl ReplayEngine {
+    pub fn new(replay: &Replay) -> Self {
+        let timing_controller = match (replay.speed(), replay.ratelimit()) {
+            (Some(speed), None) => TimingController::Speed(SpeedController::new(speed)),
+            (None, Some(ratelimit)) => TimingController::Rate(RateController::new(ratelimit)),
+            (Some(_), Some(_)) => {
+                eprintln!("speed and ratelimit cannot be specified at the same time");
+                std::process::exit(1);
+            }
+            (None, None) => {
+                eprintln!("neither speed or ratelimit are specified, defaulting to realtime speed");
+                TimingController::Speed(SpeedController::new(1.0))
+            }
+        };
+        Self {
+            command_log_path: replay.command_log().clone(),
+            timing_controller,
+        }
+    }
+
+    pub async fn generate(&mut self, replay_sender: &Sender<ClientWorkItemKind<ClientRequest>>) {
+        let command_log =
+            std::fs::File::open(&self.command_log_path).expect("failed to open command log file");
+        let reader = BufReader::new(command_log);
+
+        let mut lines_iterator = reader.lines().map_while(Result::ok);
+        if let Some(first_line) = lines_iterator.next() {
+            // Read first line to get initial timestamp first
+            let first_command = CommandLogLine::from_str(&first_line).unwrap();
+            let first_timestamp = first_command.timestamp();
+
+            // Execute command in first line
+            if replay_sender
+                .send(self.generate_cache_request(&first_command))
+                .await
+                .is_err()
+            {
+                eprintln!("Failed to send first command");
+                std::process::exit(1);
+            }
+
+            // While next lines are within 100 microseconds of the current timestamp, collect and
+            // send them out as a batch to avoid unnecessary delays and "falling behind" warnings
+            let mut current_timestamp = first_timestamp;
+            let mut batch_commands = Vec::new();
+
+            for line in lines_iterator {
+                let command = CommandLogLine::from_str(&line).unwrap();
+                if command.timestamp() - current_timestamp < TimeDelta::microseconds(100) {
+                    batch_commands.push(command);
+                } else {
+                    // Apply delay as needed to match the specified speed or rate of replay
+                    let delay_time = command.timestamp() - first_timestamp;
```

Comment:
I think we might sleep too long here since we don't account for the time to prepare all the requests in the batch?